block-storage/src/main/scala/coop/rchain/blockstorage/dag/BlockDagKeyValueStorage.scala

```diff
@@ -133,57 +135,109 @@ final class BlockDagKeyValueStorage[F[_]: Concurrent: Log] private (
   def insert(
       block: BlockMessage,
       invalid: Boolean
-  ): F[BlockDagRepresentation[F]] =
-    lock.withPermit(
+  ): F[BlockDagRepresentation[F]] = {
+    import cats.instances.list._
+    import coop.rchain.catscontrib.Catscontrib.ToBooleanF
+
+    val logAlreadyStored =
+      Log[F].warn(s"Block ${PrettyPrinter.buildString(block, short = true)} is already stored.")
+
+    def doInsert(blockMetadata: BlockMetadata): F[Unit] = {
+      val updateLatestMessages = {
+        val sendersNewLM  = (block.sender, block.blockHash)
+        val senderIsEmpty = block.sender == ByteString.EMPTY // Genesis has empty sender
+        val logEmptySender =
+          Log[F].warn(s"Block ${Base16.encode(block.blockHash.toByteArray)} sender is empty")
+
+        /** For newly bonded validators there are several options. In general, we should be as lazy as possible.
+          * And probably should add LM only when newly bonded validator creates a block.
+          * But for now we set the block where validator has been bonded as its first Latest Message.
+          * This logic is subject to some research. **/
+        // If validator is bonded but does not have self-justification in block, this is newly bonded validator
+        val newlyBondedSet = bonds(block)
+          .map(_.validator)
+          .toSet
+          .diff(block.justifications.map(_.validator).toSet)
+
+        for {
+          // This filter is required to enable adding blocks backward from higher height to lower
+          newlyBondedUnseen <- newlyBondedSet.toList.filterA(
+                                lm => latestMessagesIndex.contains(lm).not
+                              )
+          newlyBondedLMs = newlyBondedUnseen.map(v => (v, block.blockHash)).toSet
+          // As genesis has empty sender, in the very first block sender will be in newlyBondedUnseen
+          // (no self justification). In this case LM should be updated. It interferes with protection from
+          // regressive seq number for LM, so needs some additional logic to check.
+          sendersLmIsGenesis <- for {
+                                 sendersCurLM <- latestMessagesIndex.get(block.sender)
+                                 r <- sendersCurLM match {
+                                       case None => false.pure[F]
+                                       case Some(lm) =>
+                                         blockMetadataIndex.getUnsafe(lm).map(_.seqNum).map(_ == 0)
+                                     }
+                               } yield r
+
+          // Do not create LM for empty sender, which is a now valid sender for genesis
+          lms <- if (senderIsEmpty || (newlyBondedSet.contains(block.sender) && !sendersLmIsGenesis))
+                  logEmptySender.whenA(senderIsEmpty) >> newlyBondedLMs.pure[F]
```

Comment:
Log will be printed that sender is empty even when sender is not empty, and right expression is true.

Reply:
I guess you was looking at outdated code?